pkg/blockchain/noncemanager/redis/manager.go

```diff
@@ -0,0 +1,276 @@
+	nonce, err := r.cleanupAndReserveNonce(ctx)
+	if err != nil {
+		r.releaseLimiter()
+		return nil, err
+	}
+
+	metrics.EmitPayerCurrentNonce(float64(nonce))
+
+	return r.createNonceContext(nonce), nil
+}
+
+// Replenish ensures a sufficient number of nonces are available starting from the given nonce.
+// It generates up to 10,000 nonces in a single batch operation using ZADD.
+func (r *RedisBackedNonceManager) Replenish(ctx context.Context, nonce big.Int) error {
+	startNonce := nonce.Int64()
+
+	// Prepare the nonces to add
+	members := make([]redis.Z, BatchSize)
+	for i := int64(0); i < BatchSize; i++ {
+		nonceVal := startNonce + i
+		members[i] = redis.Z{
+			Score:  float64(nonceVal),
+			Member: nonceVal,
+		}
+	}
+
+	// Use a pipeline for efficiency
+	pipe := r.client.Pipeline()
+	pipe.ZAdd(ctx, r.availableKey(), members...)
+	pipe.Set(ctx, r.sequenceKey(), startNonce+BatchSize-1, 0)
+
+	if _, err := pipe.Exec(ctx); err != nil {
+		return fmt.Errorf("failed to replenish nonces in Redis: %w", err)
+	}
+
+	r.logger.Debug(
+		"Replenished nonces...",
+		zap.Uint64("starting_nonce", nonce.Uint64()),
+		zap.Int32("num_nonces", BatchSize),
+	)
+
+	return nil
+}
+
+// FastForwardNonce sets the nonce sequence to start from the given value and removes
+// all nonces below it. This is typically used when recovering from blockchain state issues.
+func (r *RedisBackedNonceManager) FastForwardNonce(ctx context.Context, nonce big.Int) error {
+	// First replenish nonces starting from the given value
+	err := r.Replenish(ctx, nonce)
+	if err != nil {
+		return err
+	}
+
+	// Remove all obsolete nonces below the given threshold
+	_, err = r.client.ZRemRangeByScore(ctx, r.availableKey(), "-inf", fmt.Sprintf("(%d", nonce.Int64())).
+		Result()
+	if err != nil {
+		return fmt.Errorf("failed to remove obsolete nonces from Redis: %w", err)
+	}
+
+	return nil
+}
+
+// cleanupAndReserveNonce atomically cleans up stale reservations and reserves the next available nonce
+func (r *RedisBackedNonceManager) cleanupAndReserveNonce(ctx context.Context) (int64, error) {
+	staleThreshold := float64(time.Now().Add(-StaleReservationTimeout).Unix())
+	currentTime := float64(time.Now().Unix())
+
+	result, err := r.client.Eval(ctx, luaCleanupAndReserveScript, []string{r.availableKey(), r.reservedKey()}, staleThreshold, currentTime).
+		Result()
+	if err != nil {
+		return 0, fmt.Errorf("failed to cleanup and reserve nonce from Redis: %w", err)
+	}
+
+	resultArray, ok := result.([]any)
+	if !ok {
+		return 0, fmt.Errorf("invalid result format from cleanup and reserve script")
+	}
+
+	// Check if a nonce was reserved (first element)
+	if len(resultArray) == 0 || resultArray[0] == nil {
+		return 0, fmt.Errorf("no nonces available in Redis")
+	}
+
+	nonceStr, ok := resultArray[0].(string)
+	if !ok {
+		return 0, fmt.Errorf("invalid nonce format from Redis")
+	}
+
+	var nonce int64
+	if _, err := fmt.Sscanf(nonceStr, "%d", &nonce); err != nil {
```

Comment:
IIRC `strconv.ParseInt` is 30X more performant than `Sscanf`, and returns the `int64` directly.

Reply:
Nice. Will swap